src/s3/util/fileViewerManager.ts

```diff
@@ -2,30 +2,198 @@
+import { downloadWithProgress } from '../commands/downloadFileAs'
+import { S3FileNode } from '../explorer/s3FileNode'
+import { readablePath } from '../util'
+import { getStringHash } from '../../shared/utilities/textUtilities'
+
+const fs = require('fs')
+const SIZE_LIMIT = 4 * Math.pow(10, 6)
+
 export class S3FileViewerManager {
-    //private cache: Set<S3FileNode>
+    private cacheArns: Set<string>
     //private activeTabs: Set<S3Tab>
-    //private window: Window
+    private window: typeof vscode.window
     private outputChannel: OutputChannel
+    private commands: Commands
+    private tempLocation: string | undefined //TODOD: create temp
 
-    public constructor(window: Window = Window.vscode(), outputChannel = ext.outputChannel) {
-        //this.cache = new Set<S3FileNode>()
+    public constructor(
+        cacheArn: Set<string> = new Set<string>(),
+        window: typeof vscode.window = vscode.window,
+        commands = Commands.vscode(),
+        tempLocation?: string
+    ) {
+        this.cacheArns = cacheArn
         //this.activeTabs = new Set<S3Tab>()
-        //this.window = window
-        this.outputChannel = outputChannel
-        //this.createTemp()
-        showOutputMessage('initializing manager', outputChannel)
+        this.window = window
+        this.commands = commands
+        this.tempLocation = tempLocation
+        this.outputChannel = ext.outputChannel
     }
 
     public async openTab(fileNode: S3FileNode): Promise<void> {
-        showOutputMessage(`     manager initialized, file: ${fileNode.file.key}`, this.outputChannel)
+        showOutputMessage(
+            `++++++++++++++++++++++++++++++++manager was initialized, file: ${fileNode.file.key}`,
+            this.outputChannel
+        )
+        const fileLocation = await this.getFile(fileNode)
+        if (!fileLocation) {
+            return
+        }
+        showOutputMessage(`file to be opened is: ${fileLocation}`, this.outputChannel)
+        //TODOD:: delegate this logic to S3Tab.ts
+        //this will display the document at the end
+        this.window.showTextDocument(fileLocation)
+    }
+
+    /**
+     * Fetches a file from S3 or gets it from the local cache if possible.
+     */
+    public async getFile(fileNode: S3FileNode): Promise<vscode.Uri | undefined> {
+        if (!this.tempLocation) {
+            await this.createTemp()
+        }
+        const targetPath = await this.createTargetPath(fileNode)
+        const targetLocation = vscode.Uri.file(targetPath)
+
+        const tempFile: vscode.Uri | undefined = await this.getFromTemp(fileNode)
+        //if it was found in temp, return the Uri location
+        if (tempFile) {
+            return tempFile
+        }
+
+        //needs to be downloaded from S3
+        //if file is +4MB, prompt user to confirm before proceeding
+        //const uri = vscode.Uri.file('')
+
+        if (fileNode.file.sizeBytes === undefined) {
```

Comment:
`if (!fileNode.file.sizeBytes) {`

Reply:
I had that before, but changed it due to a bug, should I put it back like that?

Reply:
What was the bug? Pulling files with `sizeBytes === 0`? If so, then yes, keep this (since 0 is falsy).

Reply:
`sizeBytes === 0` is _not_ the same as `sizeBytes === undefined`
This is context-specific. When the bytes are 0, this means we have received a 'folder' key and this case should be handled differently from the case of never receiving the object size.